### PR TITLE
Send service shutdown notification

### PR DESCRIPTION
**Pull request by patrickxb**
r? @maxtaco or @mmaxim

### CODE REVIEW

File: protocol/avdl/notify_service.avdl [hunk 1]
```diff
@@ -0,0 +1,7 @@
+@namespace("keybase.1")
+protocol NotifyService {
+  import idl "common.avdl";
+
+  // sent when the service shuts down.
+  void shutdown() oneway;
```
Comment (maxtaco) [inline protocol/avdl/notify_service.avdl]: you might want to wait for an ACK otherwise you can shutdown before the service has a chance to send this out!

Comment (maxtaco) [inline protocol/avdl/notify_service.avdl]: I.e., *not* `oneway`

File: go/libkb/notify_router.go [hunk 2]
```diff
@@ -363,3 +363,26 @@ func (n *NotifyRouter) HandleKeyfamilyChanged(uid keybase1.UID) {
...
+// HandleServiceShutdown is called whenever the service shuts down.
+func (n *NotifyRouter) HandleServiceShutdown() {
+	if n == nil {
+		return
+	}
+
+	n.G().Log.Debug("+ Sending service shutdown notfication")
+	// For all connections we currently have open...
+	n.cm.ApplyAll(func(id ConnectionID, xp rpc.Transporter) bool {
+		// If the connection wants the `Favorites` notification type
```
Comment (mmaxim) [inline go/libkb/notify_router.go]: Is this comment right?

File: go/libkb/notify_router.go [hunk 3]
```diff
@@ -363,3 +363,26 @@ func (n *NotifyRouter) HandleKeyfamilyChanged(uid keybase1.UID) {
...
+		if n.getNotificationChannels(id).Service {
+			// In the background do...
+			go func() {
+				(keybase1.NotifyServiceClient{
+					Cli: rpc.NewClient(xp, ErrorUnwrapper{}),
+				}).Shutdown(context.TODO())
+			}()
```
Comment (mmaxim) [inline go/libkb/notify_router.go]: Similar comment to @maxtaco, seems like the service can go down before ever sending this?

File: go/libkb/notify_router.go (hunk 2 above)
Comment (patrickxb) [inline go/libkb/notify_router.go]: urgh, copy and paste...thanks!

**Comment by patrickxb** [comment]
Ok.  How long to wait for an ACK, though?  worried service won't shut down because it's waiting for clients to acknowledge that it is shutting down.

**Comment by maxtaco** [comment]
Exactly, definitely we have to have a pretty aggressive deadline --- 5s?  @chrisnojima any specs on this?

**Comment by chrisnojima** [comment]
5s seems very long but safe. go for it

**Comment by patrickxb** [comment]
@mmaxim PTAL

File: go/libkb/notify_router.go [hunk 4]
```diff
@@ -371,18 +374,36 @@ func (n *NotifyRouter) HandleServiceShutdown() {
...
+		wg.Wait()
+		close(done)
+	}()
+
+	select {
+	case <-done:
+	case <-time.After(5 * time.Second):
+		n.G().Log.Warning("Timed out sending service shutdown notifications, proceeding to shutdown")
+	}
+
```
Comment (mmaxim) [inline go/libkb/notify_router.go]: Ah this is a cool way to implement this, nice!

**Comment by mmaxim** [comment]
👍, LGTM!

**Comment by patrickxb** [comment]
@keybase/react-hackers can you fix flow errors please?  notification protocol, added service.

**Comment by chrisnojima** [comment]
looking

**Comment by chrisnojima** [comment]
PR created: https://github.com/keybase/client/pull/3410

**Comment by chrisnojima** [comment]
Merged down

**Comment by patrickxb** [comment]
thanks

**Pull request merged**